### ISSUE TITLE
question: performance of edge queries

### ISSUE THREAD

**Issue by napei** [facebook/ent#977]
So I was digging around looking into performance of my application, when I noticed an odd set of rather expensive queries being generated that perhaps could be simplified using nested selects.

I have these schemas

<details>
<summary><code>banner</code></summary

```go
// Banner holds the schema definition for the Banner entity.
type Banner struct {
	ent.Schema
}

// Mixin defines the mixins used for this entity
func (Banner) Mixin() []ent.Mixin {
	return []ent.Mixin{
		BaseMixin{},
		mixin.Time{},
	}
}

// Fields of the Banner.
func (Banner) Fields() []ent.Field {
	return fieldutil.WithStructTags(
		field.Time("startDate"),
		field.Time("endDate"),
		field.Bool("active"),
		field.Int("order"),

		// TODO Maybe link with edges
		// External link is stored with translations
		field.String("categoryShortName").Optional(),
		field.String("merchantShortName").Optional(),
		field.String("productCode").Optional(),

		field.Bool("static1"),
		field.Bool("static2"),
	)
}

// Edges of the Banner.
func (Banner) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("translations", BannerTranslation.Type),
	}
}
```
</details>

<details>
<summary><code>bannertranslation</code></summary

```go
// BannerTranslation holds the schema definition for the BannerTranslation entity.
type BannerTranslation struct {
	ent.Schema
}

// Mixin defines the mixins used for this entity
func (BannerTranslation) Mixin() []ent.Mixin {
	return []ent.Mixin{
		BaseMixin{},
		mixin.Time{},
	}
}

// Fields of the BannerTranslation.
func (BannerTranslation) Fields() []ent.Field {
	return fieldutil.WithStructTags(
		// External link is the only link able to be overridden per language
		// as shortcodes (see banner) are independent of language
		field.String("externalLink").Optional(),
		field.String("languageCode"),
		field.Bool("active"),
	)
}

// Edges of the BannerTranslation.
func (BannerTranslation) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("banner", Banner.Type).Ref("translations").Unique().Required(),
		edge.To("image", Image.Type).Unique().Required(),
	}
}
```
</details>

<details>
<summary><code>image</code></summary

```go

// Image holds the schema definition for the Image entity.
type Image struct {
	ent.Schema
}

// Mixin defines the mixins used for this entity
func (Image) Mixin() []ent.Mixin {
	return []ent.Mixin{
		BaseMixin{},
		mixin.Time{},
	}
}

// Fields of the Image.
func (Image) Fields() []ent.Field {
	return fieldutil.WithStructTags(
		field.String("url"),
	)
}

// Edges of the Image.
func (Image) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("producttranslations", ProductTranslation.Type).Ref("images"),
		edge.From("merchanttranslation", MerchantTranslation.Type).Ref("image"),
		edge.From("categorytranslation", CategoryTranslation.Type).Ref("image"),
		edge.From("bannertranslation", BannerTranslation.Type).Ref("image"),
	}
}
```
</details>

Essentially a banner (just metadata) which mas many translations that contain language specific metadata and an image that ultimately represents the display language for that banner. 

To sum it up, what I want is `select all the banners and all the translations for each banner that match a language code and then the image for that` to be used with an endpoint like `GET banners?lang=en`

All that is fine, and I have achieved this by using this:
 
```go
	now := time.Now()
	banners, err := client.Banner.Query().
		Where(banner.Active(true)).
		Where(banner.StartDateLTE(now)).
		Where(banner.EndDateGTE(now)).
		Order(ent.Desc(banner.FieldOrder)).
		WithTranslations(func(q *ent.BannerTranslationQuery) {
			q.WithImage()
			if query.Languages != nil && len(query.Languages) > 0 {
				q.Where(bannertranslation.LanguageCodeEQ(query.Languages[0]))
			}
		}).
		All(globalContext)
```

But. 

It generates three queries, all executed sequentially, and all rather expensive.

<details>
  <summary><code>Query 1</code></summary>

```sql
SELECT DISTINCT
  "banners"."id",
  "banners"."create_time",
  "banners"."update_time",
  "banners"."start_date",
  "banners"."end_date",
  "banners"."active",
  "banners"."ORDER",
  "banners"."category_short_name",
  "banners"."merchant_short_name",
  "banners"."product_code",
  "banners"."static1",
  "banners"."static2" 
FROM
  "banners" 
WHERE
  (
    "banners"."active" = $1 
    AND "banners"."start_date" <= $2
  )
  AND "banners"."end_date" >= $3 
ORDER BY
  "ORDER" DESC
```

```
args=[true 2020-11-24 03:06:42.6622947 +0000 UTC m=+14.522046901 2020-11-24 03:06:42.6622947 +0000 UTC m=+14.522046901]
```
</details>

<details>
  <summary><code>Query 2</code></summary>

```sql
SELECT DISTINCT
  "banner_translations"."id",
  "banner_translations"."create_time",
  "banner_translations"."update_time",
  "banner_translations"."external_link",
  "banner_translations"."language_code",
  "banner_translations"."active",
  "banner_translations"."banner_translations",
  "banner_translations"."banner_translation_image" 
FROM
  "banner_translations" 
WHERE
  "banner_translations"."language_code" IN 
  (
    $1
  )
  AND "banner_translations" IN 
  (
    $2,
    $3,
    $4,
    $5,
    $6,
    $7,
    $8,
    $9,
    $10,
    $11,
    $12,
    $13,
    $14,
    $15,
    $16,
    $17,
    $18,
    $19,
    $20
  )
```

```
args=[en 03477c17-d2fc-4bdd-9e02-c9c83096d876 089eacc9-9743-43c2-aee9-e50ea7ec2e03 1a9c5db3-6e75-4615-9675-a7648dec9fbd 1e6fb10f-488d-4076-b9da-c13293752ca0 3d265b58-4a98-40c8-9f9c-cb93c266429d 480024eb-6d5d-4222-8873-131d38483d2c 538d50a5-7e8d-41ab-98fb-6557e7d8b3d9 7b8fbc76-9964-4775-9a19-958eafbdcf7d 7be7b84c-bb2c-4bcd-8914-c915a13bc772 7eeebaf1-9f97-4775-aa91-5796bb2e92f5 85e4650e-1906-4bf9-a813-a647318e2de2 922f6443-685e-4542-94a4-4b951db491eb 9f5ac742-5552-4a00-8670-c8f85cba61eb aca80fff-f30e-4c00-8eed-f9d29686362a b4be734c-f42c-40a2-9e7b-11b8368f56c2 b6143e03-a3b1-4352-8194-75b28672cf86 beb1e701-62f3-47d3-abd6-0403d20be738 e1d3c518-a84a-4ac0-b52a-a44a37f7dce7 f89e0bfc-ff29-45e7-9cc3-30a39fa3fafc]
```
</details>

<details>
  <summary><code>Query 3</code></summary>

```sql
SELECT DISTINCT
  "images"."id",
  "images"."create_time",
  "images"."update_time",
  "images"."url" 
FROM
  "images" 
WHERE
  "images"."id" IN 
  (
    $1,
    $2,
    $3,
    $4,
    $5,
    $6,
    $7,
    $8,
    $9,
    $10,
    $11,
    $12,
    $13,
    $14,
    $15,
    $16,
    $17,
    $18,
    $19
  )
```

```
args=[9d6a9d1a-e81d-463c-83f6-9feb836fa85d acc38a67-f41f-4429-90a5-b864ff5c6e8f 70919166-5484-44cc-b889-6796ab4ba2fe 72fe8711-85d0-48ff-bd8c-6ad60fab5b40 edea7c31-edb4-4723-925e-ddceeeff2b71 e06224e8-14f5-4045-a2f6-ab55acf29280 0ef74a48-df4b-4825-8666-394d28b631fe 470a9ba2-b560-47cc-826c-df52e06de8e2 56e38abd-1706-478c-a635-10715a459a85 78c19471-4cc3-410d-a23d-b488d2e431a3 101b3d09-b2d3-42a1-8dca-24bd394b07a0 458f5630-3db4-4ff2-8a07-b57366718ed6 ca68aed3-0296-4903-b7c9-55ae9150ce8e 3a76737d-0f8c-4ef6-a0b1-3f7d623be352 85a8e96d-8125-4d80-8466-d5114df669e2 e5a14323-fbce-410e-b911-f5b0822502cc efd21b72-3b90-4a2d-9006-c76b2331887f e3372149-82e9-4807-966e-d03c0c70d9fa ff83bc05-b10d-419c-946b-52252a66e78f]
```
</details>

Raw logs for reference (timing): https://pastebin.com/raw/ghHp68EP

As you can see they are quite expensive and take some time. Specifically, they are made one after another, instead of using something like nested queries or joins. On top of that there is a lot of memory allocation happening in go (I assume) to get the result of each query and store all those UUIDs. I don't think I need to give an example of a dataset with thousands of rows and what kind of data transfer back and forth would happen.

What could I do differently in order for it to not go down this expensive path of listing out every ID it can find as args? Or, more specifically. Probably nesting the queries would be an alternative, however multiple models are being retrieved here so perhaps it's more applicable to a join?

I'm most definitely not an SQL expert though so there may be a perfectly valid reason for this, I just don't see it scaling well.

**Comment by napei** [comment]
Just as a thought experiment I tried writing a query for this. 

```sql
SELECT
  * 
FROM
  banners AS b 
  INNER JOIN
    public.banner_translations AS bt 
    ON bt.banner_translations = b.id 
    AND b.active = TRUE 
    AND bt.active = TRUE 
    AND bt.language_code = 'en' 
  INNER JOIN
    images AS im 
    ON im.id = bt.banner_translation_image
```

**Comment by a8m** [comment]
Hey @napei and thanks for raising up this issue!

Regarding your question, as mentioned in the [docs](https://entgo.io/docs/eager-load/#implementation): <i>since a query-builder can load more than one association, it's not possible to load them using one `JOIN` operation. Therefore, ent executes additional queries for loading associations. One query for M2O/O2M and O2O edges, and 2 queries for loading M2M edges.</i>

In some cases, executing multiple queries reduces the number of rows returned from the database and the total time of the query - even if there are multiple roundtrips. 

One optimization I want to add is to use `JOIN` when loading only one association (edge), or at least give users an SQL specific API to tell ent how to load the edges (e.g. use `JOIN` even if the query builder contains more than 2 association).

> As you can see they are quite expensive and take some time. Specifically, they are made one after another, instead of using something like nested queries or joins

Can you please elaborate what do you mean by "nested queries"? how do you query multiple tables with "nested queries"?

**Comment by napei** [comment]
> Can you please elaborate what do you mean by "nested queries"? how do you query multiple tables with "nested queries"?

Yeah so instead of generating `n` arguments for the second query and wasting all that space, it might be more efficient both memory and data wise to use a subquery?

There are still 3 queries in this example one for each but the number of parameters is cut down significantly and a well indexed database might be able to execute these fairly well.

For example, when querying `banner_translations` in my example 
```sql
SELECT
  * 
FROM
  banner_translations 
WHERE
  "banner_translations" IN 
  (
    SELECT
      "id" 
    FROM
      banners 
    WHERE
      active = TRUE
      ...etc
  )
```

and then to get images

```sql
SELECT
  * 
FROM
  images 
WHERE
  "id" IN 
  (
    SELECT
      "banner_translation_image" 
    FROM
      banner_translations 
    WHERE
      "banner_translations" IN 
      (
        SELECT
          "id" 
        FROM
          banners 
        WHERE
          active = TRUE
      )
  )
```

Now, I don't know specifically if this query is actually better **performance** or not but it's definintley more efficient than having `n` parameters for however many of the original record is found that has to be both dealt with in code and transmitted over the network.

**Comment by a8m** [comment]
> Yeah so instead of generating n arguments for the second query and wasting all that space, it might be more efficient both memory and data wise to use a subquery?

In some cases the root-query is a traversal of `N` steps with a few filters, (e.g. `t.QueryT1().Where(...).QueryT2().WithT3()`), and I'm pretty sure it will be much less efficient to execute this sub-query on each loaded-relation - I think the current approach is faster for most cases, but you're welcome to help with testing it and give me more insights if you think it can be improved.


> Now, I don't know specifically if this query is actually better performance or not but it's definitely more efficient than having n parameters for however many of the original record is found that has to be both dealt with in code and transmitted over the network.

I think we care more about the database performance rather than the application performance, right?

**Comment by napei** [comment]
So I did some testing. I added 10,000 rows of banners, banner_translations, and images and tested the performance. I parsed the logs given by `ent` in debug mode to get the query used with argument substitution and ran them in pgAdmin on my local machine (Melbourne, Australia) connected to a postgres server in GCP Singapore. It's just a small 1 vCPU & 3.75GB RAM server.

This isn't exactly an in-depth test, the relationship between these entities is very linear, and the data is very simple. I applied no where clauses to the search either. The goal is to load a banner, its translation(s) and their image. Each query was run 5 times and the last result taken to ensure the database had "warmed up".

### Ent results

According to `go`, selecting all banners as follows took 3.78 seconds, which is pretty good but still a bit slow for only 10,000 records

```go
_, err = client.Banner.Query().WithTranslations(func(btq *ent.BannerTranslationQuery) {
  btq.WithImage()
}).All(ctx)
if err != nil {
  log.Fatalln(err)
}
log.Println("duration: ", time.Since(start).String())
```

<details>
<summary>Banners ~ 12ms</summary>

![image](https://user-images.githubusercontent.com/8946502/100348654-4beda000-303b-11eb-9358-fb7676a5cd75.png)
</details>

<details>
<summary>Banner Translations ~280ms</summary>

![image](https://user-images.githubusercontent.com/8946502/100348707-632c8d80-303b-11eb-9821-32c061c95652.png)
</details>

<details>
<summary>Images ~ 30ms</summary>

![image](https://user-images.githubusercontent.com/8946502/100348751-77708a80-303b-11eb-8d1d-6698c2a23fbb.png)
</details>


### Subquery Results
The one that takes the longest is the banner translations. Surprise suprise I had no indexes as an oversight (still in development) on it however the subquery performs almost the same as the others even with no index present.

<details>
<summary>Banner Translations (subquery, no indexes) ~ 21ms</summary>

![image](https://user-images.githubusercontent.com/8946502/100348900-b8689f00-303b-11eb-977c-8f705a71930c.png)
</details>

Finally after adding indexes the `ent` generated query for `banner_translations` runs well now.

<details>
<summary>Banner Translations (ent, with indexes) ~ 28ms</summary>

![image](https://user-images.githubusercontent.com/8946502/100350247-b1429080-303d-11eb-85aa-8404aeccaba5.png)
</details>

Interestingly enough, the subquery version of images runs just as fast even without indexes on the intermediate `banner_translations` table

<details>
<summary>Images (subquery, no index on banner_translations) ~ 30ms</summary>

![image](https://user-images.githubusercontent.com/8946502/100351151-164ab600-303f-11eb-9af7-f2182cea1395.png)
</details>

So I think, there is a line where subqueries are faster in general just because maybe the database is optimised to complete them faster. Applying them everywhere may not always provide the performance benefits here but at least for this particular example there is.

I did see an improvement just adding indexes to the data, however that isn't always the solution; this is just one example.

I don't know how this applies to much more deeply-nested models, and I don't really know if the results I got are good or bad. 

Some form of optimisation like this would probably be good to have as an opt-in feature or something like that. Perhaps even a generation-time analysis of the complexity of the schemas and how they could benefit from either or based on some assumptions. Just throwing ideas around.

Either way, I'm happy for queries to be left how they are and I just get better at remembering to index my schemas :joy:


TL;DR subquerys are **sometimes** faster, but there are many cases (probably) that they may not be.

**Comment by a8m** [comment]
@napei - Wow! I don't know how I missed this comment, but thanks a lot running these tests and sharing with me the insights. I assume it took you some time to do it, and it's really appreciated - that's one of the reasons I really like the OSS community 😍 

---

> Some form of optimisation like this would probably be good to have as an opt-in feature or something like that. Perhaps even a generation-time analysis of the complexity of the schemas and how they could benefit from either or based on some assumptions. Just throwing ideas around.

That sounds like a nice idea, but I'm a little bit afraid that it'll add another complexity to the framework - If you're really into it, let's start with throwing some ideas on how such API should look like and then I'll have a better sense on the implementation complexity. 

> My biggest worry/annoyance was that 10,000 x 2 UUID were getting parsed, saved, and sent along the wire (~720 kB provided they're sent in plain text (36 bytes each)) which in the modern world isn't that much but it also isn't small just for the sake of a query.

Yeah, in cases of querying 10K objects with their relations (recursively), I guess the suggestion above can be useful. 

> Either way, I'm happy for queries to be left how they are and I just get better at remembering to index my schemas 😂

😃

**Comment by a8m** [comment]
@napei - Wow! I don't know how I missed this comment, thanks a lot for running these tests and sharing the insights. I assume it took you some time to do it, and it's really appreciated - that's one of the reasons I really like the OSS community 😍 

---

> Some form of optimisation like this would probably be good to have as an opt-in feature or something like that. Perhaps even a generation-time analysis of the complexity of the schemas and how they could benefit from either or based on some assumptions. Just throwing ideas around.

That sounds like a nice idea, but I'm a little bit afraid that it'll add another complexity to the framework - If you're really into it, let's start with throwing some ideas on how such API should look like and then I'll have a better sense on the implementation complexity. 

> My biggest worry/annoyance was that 10,000 x 2 UUID were getting parsed, saved, and sent along the wire (~720 kB provided they're sent in plain text (36 bytes each)) which in the modern world isn't that much but it also isn't small just for the sake of a query.

Yeah, in cases of querying 10K objects with their relations (recursively), I guess the suggestion above can be useful. 

> Either way, I'm happy for queries to be left how they are and I just get better at remembering to index my schemas 😂

😃

**Comment by napei** [comment]
@a8m - I'm not really sure how that would work now that I came to the conclusion that indexing fixes it. Perhaps starting with a note in the docs about how queries are formed behind the scenes, and how these performance caveats can impact performance and their solution (indexing) would be good just from an information perspective.

From an implementation point of view, I could see either a per-query flag, or a configuration parameter to `ent generate` to say either use nested subqueries globally, or not.

Per-query could be something like 

```go
client.Banner.Query().WithTranslations().All(ctx, options...)
```

Where `options...` would be flags, one of which is `QueryMethodNested`

This way the default behaviour remains, but queries that would benefit from using subqueries could have this enabled. This would, of course, bloat out the generated code a bit with code needing to be present for both methods, but I suppose that's what compiler optimisations are for.

The other global option could work but it would have to be tested with every possible data structure to ensure it works, something I haven't done yet.

**Comment by a8m** [comment]
> @a8m - I'm not really sure how that would work now that I came to the conclusion that indexing fixes it. Perhaps starting with a note in the docs about how queries are formed behind the scenes, and how these performance caveats can impact performance and their solution (indexing) would be good just from an information perspective.

What special indexing did you create on your database? FKs should already be indexed by default AFAIK. 

> From an implementation point of view, I could see either a per-query flag, or a configuration parameter to ent generate to say either use nested subqueries globally, or not.

Maybe add an option to add annotation on the edge (in the schema), and then expose functionality on the query-builder to be used by eager-loading as follows:

```go
client.Pet.
    Query().
    WithFriends(func (q *ent.PetQuery) {
        q.WithOptions(opts...)
    })
    All(ctx)
```

**Comment by napei** [comment]
> What special indexing did you create on your database? FKs should already be indexed by default AFAIK. 

I don't have that code anymore, but from memory I stuck indexes on the edges and the primary keys. 

> Maybe add an option to add annotation on the edge (in the schema), and then expose functionality on the query-builder to be used by eager-loading as follows:

That would work I think, but maybe `WithOptions` is a little confusing as that's another reserved word that a schema can't be called to add to the list (unless it already is one). Another way could be a parameter in the `WithFriends` function to take the options as `opts...`? Like `hook.On`
